capsules/src/sip_hash.rs

```diff
@@ -0,0 +1,287 @@
+//! Tock SipHash capsule.
+//!
+//! This is a async implementation of the SipHash.
+//!
+//! This capsule was originally written to be used as part of Tocks
+//! key/value store. SipHash was used as it is generally fast, while also
+//! being resilient against DOS attacks from userspace
+//! (unlike <https://github.com/servo/rust-fnv>).
+//!
+//! Read <https://github.com/veorq/SipHash/blob/master/README.md> for more
+//! details on SipHash.
+//!
+//! The implementation is based on the Rust implementation from
+//! rust-core, avaliable here: <https://github.com/jedisct1/rust-siphash>
+//!
+//! Copyright 2012-2016 The Rust Project Developers.
+//! Copyright 2016-2021 Frank Denis.
+//! Copyright 2021 Western Digital
+//!
+//! Licensed under the Apache License, Version 2.0 LICENSE-APACHE or
+//! <http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+//! LICENSE-MIT or <http://opensource.org/licenses/MIT>, at your
+//! option.
+
+use core::cell::Cell;
+use core::convert::TryInto;
+use core::{cmp, mem};
+use kernel::dynamic_deferred_call::{
+    DeferredCallHandle, DynamicDeferredCall, DynamicDeferredCallClient,
+};
+use kernel::hil::hasher::{Client, Hasher};
+use kernel::utilities::cells::{OptionalCell, TakeCell};
+use kernel::utilities::leasable_buffer::LeasableBuffer;
+use kernel::ErrorCode;
+
+pub struct SipHasher24<'a> {
+    client: OptionalCell<&'a dyn Client<'a, 8>>,
+
+    hasher: Cell<SipHasher>,
+
+    add_data_deferred_call: Cell<bool>,
+    complete_deferred_call: Cell<bool>,
+    deferred_caller: &'static DynamicDeferredCall,
+    deferred_handle: OptionalCell<DeferredCallHandle>,
+
+    data_buffer: TakeCell<'static, [u8]>,
+    out_buffer: TakeCell<'static, [u8; 8]>,
+}
+
+#[derive(Debug, Clone, Copy)]
+struct SipHasher {
+    k0: u64,
+    k1: u64,
+    length: usize, // how many bytes we've processed
+    state: State,  // hash State
+    tail: u64,     // unprocessed bytes le
+    ntail: usize,  // how many bytes in tail are valid
+}
+
+#[derive(Debug, Clone, Copy)]
+struct State {
+    // v0, v2 and v1, v3 show up in pairs in the algorithm,
+    // and simd implementations of SipHash will use vectors
+    // of v02 and v13. By placing them in this order in the struct,
+    // the compiler can pick up on just a few simd optimizations by itself.
+    v0: u64,
+    v2: u64,
+    v1: u64,
+    v3: u64,
+}
+
+impl<'a> SipHasher24<'a> {
+    pub const fn new(deferred_caller: &'static DynamicDeferredCall) -> Self {
+        let hasher = SipHasher {
+            k0: 0,
```

Comment:
There seems to be no way to set `k0` or `k1`?

Reply:
No, there isn't. We have no need to set the keys at this point. In the future if we do we can add a `new_with_keys()` function that sets them

Reply:
If you're going to implement SipHash, please implement it. :) If you want a partial implementation that does just what your use case needs, that's fine, but then it should be in the TickV capsule and not a general capsule that claims to implement a particular PRF.

I don't think you want a `new_with_keys()`, because this would make the key fixed for the lifetime of the PRF, which would make it not reusable. Instead, shouldn't it be done as it is with SHA HMACs in Digest, where there's a way to set it to a mode, and setting the mode passes the key? I think you want to structure this almost identically to `Digest`, it just has different computational properties, kind of like how `Entropy` and `Random` are structurally so similar.

Reply:
I can add a `new_with_keys()` then.

As for changing the keys, I don't think that's something we want to support. The original Rust implementation doesn't: https://github.com/jedisct1/rust-siphash/blob/master/src/sip.rs#L207

I also found some other implementations and they don't support changing the keys on the fly. If we want new keys we can just create a new instance

Reply:
As for structuring this like `Digest`, I don't think we want to do that. `Digest` is aimed at something that supports SHA and we can then specify different SHA-like operations.

This on the other hand is much more specilised. The idea is someone creates a `SipHash` instance and then uses that. This is more similar to Rusts Hasher trait then Digest

Reply:
> I can add a `new_with_keys()` then.
> 
> As for changing the keys, I don't think that's something we want to support. The original Rust implementation doesn't: https://github.com/jedisct1/rust-siphash/blob/master/src/sip.rs#L207
> 
> I also found some other implementations and they don't support changing the keys on the fly. If we want new keys we can just create a new instance

But we don't have dynamic allocation. If we had dynamic allocation, then a static key at initialization is fine, because you can create new instances if new keys appear.

Reply:
> As for structuring this like `Digest`, I don't think we want to do that. `Digest` is aimed at something that supports SHA and we can then specify different SHA-like operations.
> 
> This on the other hand is much more specilised. The idea is someone creates a `SipHash` instance and then uses that. This is more similar to Rusts Hasher trait then Digest

I'm talking about the HIL. `Digest` provides cryptographic hashes, which have particular mathematical properties. `Hasher` (which should be `Hash`) provides non-cryptographic hashes, which have other mathematical properties. If you just want to provide a specialized `SipHash`, that's fine, there just shouldn't be a HIL around it.

I agree that `Hash` wants to be a bit different from `Digest`, in particular we don't need to differentiate verification and computation. But overall its structure should be similar.